### PR TITLE
[Policy Insights] Added cmdlets for policy attestations

### PR DESCRIPTION
**Pull request by samarthlakhotia**
## Description

This PR is for introducing Policy Attestations in Azure. 

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+using System.Linq;
+using System.Management.Automation;
+using System.Text;
+
+namespace Microsoft.Azure.Commands.PolicyInsights.Cmdlets.Attestations
+{
+    /// <summary>
+    /// Updates a policy attestation.
+    /// </summary>
+    [Cmdlet(verbName: VerbsCommon.Set, AzureRMConstants.AzureRMPrefix + "PolicyAttestation", DefaultParameterSetName = ParameterSetNames.ByName, SupportsShouldProcess = true), OutputType(typeof(PSAttestation))]
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs:36]: Is this a Patch API or a Put API? If it's a Patch, please use update as the verb of this cmdlet.

File: src/PolicyInsights/PolicyInsights/help/Az.PolicyInsightsCmdlet.Design.md [hunk 2]
```diff
@@ -0,0 +1,1283 @@
...
+Stop-AzPolicyRemediation -ResourceId <String> [-AsJob] [-PassThru] [-DefaultProfile <IAzureContextContainer>]
+ [-WhatIf] [-Confirm] [<CommonParameters>]
+```
+
++ ByInputObject
+```powershell
+Stop-AzPolicyRemediation -InputObject <PSRemediation> [-AsJob] [-PassThru]
+ [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm] [<CommonParameters>]
+```
+
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/Az.PolicyInsightsCmdlet.Design.md:825]: This doc should not be checked in

File: src/PolicyInsights/PolicyInsights/help/Get-AzPolicyAttestation.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+## DESCRIPTION
+The **Get-AzPolicyAttestation** cmdlet gets all policy attestations in a scope or a particular attestation.
+
+## EXAMPLES
+
+### Example 1: Get all policy attestations in the current subscription
+```powershell
+Set-AzContext -Subscription "MySubscription"
+Get-AzPolicyAttestation
+```
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/Get-AzPolicyAttestation.md:53]: please add output section for each example
\```output
{output}
\```

File: src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+Creates a new policy attestation for a policy assignment.
+
+## SYNTAX
+
+### ByName (Default)
+```
+New-AzPolicyAttestation -Name <String> [-Scope <String>] [-ResourceGroupName <String>]
+ -PolicyAssignmentId <String> [-ComplianceState <String>] [-PolicyDefinitionReferenceId <String>]
+ [-ExpiresOn <DateTime>] [-Owner <String>] [-Comment <String>] [-Evidence <PSAttestationEvidence[]>]
+ [-AssessmentDate <DateTime>] [-Metadata <Object>] [-DefaultProfile <IAzureContextContainer>] [-WhatIf]
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md:20]: What's the format for Metadata value? Object is an obscure object type

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

**Review by eladperets — DISMISSED** [review]
LGTM!

File: src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs (hunk 1 above)
Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs:36]: This is a PUT Api

File: src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md (hunk 4 above)
Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md:20]: @BethanyZhou The format for Metadata is JObject (JSON Object). I have wrapped it as a PSAttestationEvidence object. I will regerate this file to show the updated value. Thank you

Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md:20]: I will wrap it with PSAttestationMetadata type

File: src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        [Parameter(ParameterSetName = ParameterSetNames.ByResourceId, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = ParameterHelpMessages.ResourceId)]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        [Parameter(ParameterSetName = ParameterSetNames.ByInputObject, Mandatory = true, ValueFromPipeline = true, HelpMessage = ParameterHelpMessages.AttestationObject)]
+        [ValidateNotNull]
+        public PSAttestation InputObject { get; set; }
+
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = ParameterHelpMessages.PolicyAssignmentId)]
+        public string PolicyAssignmentId { get; set; }
```
Comment (pilor) [inline src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs:62]: >PolicyAssignmentId [](http://example.com/codeflow?start=22&length=18)

make sure all of these properties can actually be updated on an existing attestation

File: src/PolicyInsights/PolicyInsights/Common/AttestationCmdletBase.cs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        protected string GetAttestationName(string name = null, string resourceId = null, PSAttestation inputObject = null)
+        {
+            string attestationName = null;
+            if (!string.IsNullOrEmpty(name))
+            {
+                attestationName = name;
+            }
+            else if (!string.IsNullOrEmpty(resourceId))
+            {
+                attestationName = ResourceIdHelper.GetResourceName(resourceId: resourceId);
```
Comment (pilor) [inline src/PolicyInsights/PolicyInsights/Common/AttestationCmdletBase.cs:88]: > attestationName = ResourceIdHelper.GetResourceName(resourceId: resourceId); [](http://example.com/codeflow?start=15&length=76)

make sure this actually works if the attestation is extending another resource (i.e. [https://learn.microsoft.com/en-us/rest/api/policy/attestations/create-or-update-at-resource?tabs=HTTP](https://learn.microsoft.com/en-us/rest/api/policy/attestations/create-or-update-at-resource?tabs=HTTP))

File: src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs (hunk 5 above)
Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/Cmdlets/Attestations/SetAzureRmPolicyAttestation.cs:62]: @pilor iirc policyassignmentid cannot be changed for an existing attestation.

**Comment by samarthlakhotia** [comment]
@BethanyZhou @isra-fel I have addressed all the above PR comments. Please review :) Thank you

File: src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md [hunk 7]
```diff
@@ -0,0 +1,459 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.PolicyInsights.dll-Help.xml
+Module Name: Az.PolicyInsights
+online version:
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md:4]: Please add online version for each new cmdlet docs. its pattern should be https://learn.microsoft.com/powershell/module/az.policyinsights/{lowerCase(cmdletname)}

File: src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+
+### System.Object
+
+## OUTPUTS
+
+### Microsoft.Azure.Commands.PolicyInsights.Models.Attestations.PSAttestation
+
+## NOTES
+
+## RELATED LINKS
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md:459]: Please add related cmdlet links to this section if possible

File: src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md (hunk 4 above)
Comment (BethanyZhou) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: Is it applicable to use dictionary or something else to make this interface more friend to customer?

Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: @BethanyZhou Internally (on the server side), metadata is represented as a JSON object. So as long as one passes a valid JSON string, it should work fine. What would you recommend using instead such that we can make it more user friendly? Should we expose Metadata as a dictionary type and then parse it as a json object while processing it?

Comment (BethanyZhou) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: Yes, that's what i'm thinking. what's your thought about this idea?

Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: I was thinking that I can make it a string, similar to how New-AzPolicyDefinition does it, here https://github.com/Azure/azure-powershell/blob/4668bff1cc4ed6510d9b0d4c6ec289ab27f04241/src/Resources/ResourceManager/Implementation/Policy/NewAzurePolicyDefinition.cs#L66. This parameter supports a string containing json content or a path to a json file.

Comment (BethanyZhou) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: Sounds good

Comment (samarthlakhotia) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: @BethanyZhou Made the changes. Let me know if this works. I will further add a test or two to verify passing in a json file for metadata

Comment (BethanyZhou) [reply src/PolicyInsights/PolicyInsights/help/New-AzPolicyAttestation.md]: Looks good to me, please go ahead.

File: src/PolicyInsights/PolicyInsights/Models/Attestations/PSAttestationMetadata.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.PolicyInsights.Models.Attestations
+{
+    public class PSAttestationMetadata
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/Models/Attestations/PSAttestationMetadata.cs:21]: This class seems no need anymore, please confirm and remove it if ture.

File: src/PolicyInsights/PolicyInsights/help/Get-AzPolicyAttestation.md [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+PolicyDefinitionReferenceId : pstestattestationrg_1
+ComplianceState             :
+ExpiresOn                   : 1/27/2024 2:51:54 AM
+Owner                       : Test Owner
+Comment                     :
+Evidence                    : {Microsoft.Azure.Commands.PolicyInsights.Models.Attestations.PSAttestationEvidence}
+ProvisioningState           : Succeeded
+LastComplianceStateChangeAt : 1/27/2023 2:51:57 AM
+AssessmentDate              : 1/25/2024 2:51:54 AM
+Metadata                    : Microsoft.Azure.Commands.PolicyInsights.Models.Attestations.PSAttestationMetadata
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/Get-AzPolicyAttestation.md:126]: Update example accordingly please.

File: src/PolicyInsights/PolicyInsights/help/Get-AzPolicyAttestation.md [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+### System.String
+
+## OUTPUTS
+
+### Microsoft.Azure.Commands.PolicyInsights.Models.Attestations.PSAttestation
+
+## NOTES
+
+## RELATED LINKS
+[Attestation Overview](https://learn.microsoft.com/en-us/azure/governance/policy/concepts/attestation-structure)
```
Comment (BethanyZhou) [inline src/PolicyInsights/PolicyInsights/help/Get-AzPolicyAttestation.md:336]: It's nice to add links for its related commands like: https://github.com/Azure/azure-powershell/blob/1cc694b9fba9b84fe300cec4bea335951e19182a/src/KeyVault/KeyVault/help/Get-AzKeyVault.md?plain=1#L346-L349

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

**Comment by samarthlakhotia** [comment]
@BethanyZhou I have made the suggested changes. Let me know if you have any questions. Thanks

**Comment by samarthlakhotia** [comment]
@isra-fel @BethanyZhou Can we merge this whenever you get a chance? Thank you

**Review by BethanyZhou — APPROVED** [review]
LGTM

**Comment by samarthlakhotia** [comment]
@BethanyZhou I wanted to check in to see if this was released? We do not see the documentation for this on learn.microsoft.com

**Comment by BethanyZhou** [comment]
Unfortunately we had to delay the official release for a few days. ETA end of week. Sorry for the inconvenience.

**Comment by BethanyZhou** [comment]
Az 9.5.0 was released. Please check the status of your features.

**Pull request merged**